livekit-plugins/livekit-plugins-cartesia/livekit/plugins/cartesia/tts.py

```diff
@@ -137,3 +148,176 @@ async def _main_task(self):
                         request_id=request_id, segment_id=segment_id, frame=frame
                     )
                 )
+
+class SynthesizeStream(tts.SynthesizeStream):
+    """Streamed API using websockets"""
+
+    def __init__(
+        self,
+        session: aiohttp.ClientSession,
+        opts: _TTSOptions,
+    ):
+        super().__init__()
+        self._opts, self._session = opts, session
+
+    @utils.log_exceptions(logger=logger)
+    async def _main_task(self) -> None:
+        self._segments_ch = utils.aio.Chan[tokenize.SentenceStream]()
+
+        @utils.log_exceptions(logger=logger)
+        async def _tokenize_input():
+            """tokenize text from the input_ch to sentences"""
+            sentence_stream = None
+            async for input in self._input_ch:
+                if isinstance(input, str):
+                    if sentence_stream is None:
+                        # new segment (after flush for e.g)
+                        sentence_stream = self._opts.sentence_tokenizer.stream()
+                        self._segments_ch.send_nowait(sentence_stream)
+
+                    sentence_stream.push_text(input)
+                elif isinstance(input, self._FlushSentinel):
+                    if sentence_stream is not None:
+                        sentence_stream.end_input()
+
+                    sentence_stream = None
+
+            self._segments_ch.close()
+
+        @utils.log_exceptions(logger=logger)
+        async def _run():
+            async for sentence_stream in self._segments_ch:
+                await self._run_ws(sentence_stream)
+
+        tasks = [
+            asyncio.create_task(_tokenize_input()),
+            asyncio.create_task(_run()),
+        ]
+        try:
+            await asyncio.gather(*tasks)
+        finally:
+            await utils.aio.gracefully_cancel(*tasks)
+
+    async def _run_ws(
+        self,
+        sentence_stream: tokenize.SentenceStream,
+        max_retry: int = 3,
+    ) -> None:
+        ws_conn: aiohttp.ClientWebSocketResponse | None = None
+        for try_i in range(max_retry):
+            retry_delay = 5
+            try:
+                if try_i > 0:
+                    await asyncio.sleep(retry_delay)
+
+                ws_conn = await self._session.ws_connect(
+                    f"wss://api.cartesia.ai/tts/websocket?api_key={self._opts.api_key}&cartesia_version={API_VERSION}",
+                )
+                break
+            except Exception as e:
+                logger.warning(
+                    f"failed to connect to Cartesia, retrying in {retry_delay}s",
+                    exc_info=e,
+                )
+
+        if ws_conn is None:
+            raise Exception(f"failed to connect to Cartesia after {max_retry} retries")
```

Comment:
@theomonnom this follows the pattern set by the 11labs websocket plugin, however, i'm not sure if making a new websocket connection on every `sentence_stream` or `word_stream` is the right approach because that introduces latency associated with the websocket handshake

Reply:
The reason why the 11labs connection is closed is because they don't support to make a new request in the same connection after a flush

Reply:
gotcha! we should tweak this to maintain the same WS connection and refresh it on each call.